src/circuit/circuit_frame.rs

```diff
@@ -3055,6 +3175,198 @@ fn apply_continuation<F: LurkField, CS: ConstraintSystem<F>>(
 
         let res = AllocatedPtr::from_parts(&res_tag, &val);
 
+        // Next constraints are used for number comparisons
+        ///////////////////////////////////////////////////////////////////////
+        let double_a = constraints::add(&mut cs.namespace(|| "double a"), a, a)?;
+        let double_a_bits = double_a
+            .to_bits_le_strict(&mut cs.namespace(|| "double a lsb"))
+            .unwrap();
+        let lsb_a = double_a_bits.get(0);
+        let bool_val_a = match lsb_a {
+            Some(Boolean::Is(v)) => v.get_value(),
+            _ => {
+                Some(false) // Blank
+            }
+        };
+        let a_is_negative = Boolean::from(
+            AllocatedBit::alloc(cs.namespace(|| "boolean val a"), bool_val_a).unwrap(),
+        );
+
+        let double_b = constraints::add(&mut cs.namespace(|| "double b"), b, b)?;
+        let double_b_bits = double_b
+            .to_bits_le_strict(&mut cs.namespace(|| "double b lsb"))
+            .unwrap();
+        let lsb_b = double_b_bits.get(0);
+        let bool_val_b = match lsb_b {
+            Some(Boolean::Is(v)) => v.get_value(),
+            _ => {
+                Some(false) // Blank
+            }
+        };
+        let b_is_negative = Boolean::from(
+            AllocatedBit::alloc(cs.namespace(|| "boolean val b"), bool_val_b).unwrap(),
+        );
+
+        let both_positive = Boolean::and(
+            &mut cs.namespace(|| "both positive"),
+            &Boolean::not(&a_is_negative),
+            &Boolean::not(&b_is_negative),
+        )?;
+        let both_negative = Boolean::and(
+            &mut cs.namespace(|| "both negative"),
+            &a_is_negative,
+            &b_is_negative,
+        )?;
+        let both_same_sign = constraints::or(
+            &mut cs.namespace(|| "both same sign"),
+            &both_negative,
+            &both_positive,
+        )?;
+        let a_negative_and_b_positive = Boolean::and(
+            &mut cs.namespace(|| "a negative and b positive"),
+            &a_is_negative,
+            &Boolean::not(&b_is_negative),
+        )?;
+
+        let diff_is_zero = alloc_is_zero(&mut cs.namespace(|| "diff is zero"), &diff)?;
+        let double_diff = constraints::add(&mut cs.namespace(|| "double diff"), &diff, &diff)?;
+        let double_diff_bits = double_diff.to_bits_le_strict(&mut cs).unwrap();
+        let lsb_diff = double_diff_bits.get(0);
+        let bool_val_diff = match lsb_diff {
+            Some(Boolean::Is(v)) => v.get_value(),
+            _ => {
+                Some(false) // Blank
+            }
+        };
+        let diff_is_negative = Boolean::from(
+            AllocatedBit::alloc(cs.namespace(|| "diff is negative"), bool_val_diff).unwrap(),
+        );
+
+        // LESS
+        let comp_val_less = AllocatedPtr::pick(
+            &mut cs.namespace(|| "comp val less"),
+            &diff_is_negative,
+            &g.t_ptr,
+            &g.nil_ptr,
+        )?;
```

Comment:
I don't think it's necessary or efficient to convert these booleans to `AllocatedPtr`s here. By doing so, you're adding extra case clauses, and also making all the subsequent boolean manipulations more expensive.

Instead, I think you should be converting to an `AllocatedBit` (in this case, the one allocated above). Ideally, we would be able to use an `AllocatedBit` as-is in a `CaseClause`, but this would require extending the case/multicase mechanism perhaps significantly.

We could avoid that (at least for now — though it seems the benefit might be worth doing it eventually), by converting from `AllocatedBit` to `AllocatedNum` and back (if necessary) after the case selection.

In theory, we should be able to get the conversion to `AllocatedNum` 'for free', but this would require extending `AllocatedNum` to allow a free downgrade. That is, we should be able to reuse the R1CS variable while losing the constraint. We could then convert back and accept the cost of an extra constraint to re-establish booleanity.

For maximum simplicity, we can probably still accomplish everything needed with three constraint total:
- one equality: bit = num
- one equality: num = bit
- one booleanity: bit is boolean

Once we have the final value (after all the cases and post-case boolean logic), we can convert to a Lurk value last.

We should have a gadget which allocates a Lurk boolean `AllocatedPtr` (`t` or `nil`) from a boolean (and/or `AllocatedBit`, as needed), and this should be applied only to the final expressions. We should keep boolean logic in circuit-native terms until it needs to be converted into Lurk data for output.

Reply:
I realize the above is a little confusing, and possibly overlooks some detail. Ping me directly if you want to talk through the issues. I think the main point stands and is worth consideration.